test/pino-opentelemetry-transport.test.js

```diff
@@ -0,0 +1,226 @@
+'use strict'
+
+const { mkdir, rm, readFile } = require('fs/promises')
+const { join } = require('path')
+const { test, beforeEach } = require('tap')
+const { promisify } = require('util')
+const { tmpdir } = require('os')
+const requireInject = require('require-inject')
+
+async function cleanAndCreateFolder (path) {
+  await rm(path, { force: true, recursive: true })
+  await mkdir(path, { force: true, recursive: true })
+}
+
+const sleep = promisify(setTimeout)
+
+const logFolder = join(tmpdir(), 'test-logs')
+
+beforeEach(() => cleanAndCreateFolder(logFolder))
+
+const logFile = join(logFolder, 'pino.log')
+const MOCK_HOSTNAME = 'hostname'
+
+test('translate Pino log format to Open Telemetry data format for each log level', async ({ same, afterEach }) => {
+  const ms = 1531069919686
+  const now = Date.now
+  Date.now = () => ms
+
+  afterEach(() => {
+    Date.now = now
+
+    return rm(logFolder, { force: true, recursive: true })
+  })
+
+  const pino = requireInject.withEmptyCache('pino', {
+    os: {
+      hostname: () => MOCK_HOSTNAME
+    }
+  })
+
+  const transport = pino.transport({
+    level: 'trace',
+    target: '..',
+    options: { destination: logFile }
+  })
+
+  const logger = pino(transport)
+  logger.level = 'trace'
+
+  logger.trace('test trace')
+  logger.debug('test debug')
+  logger.info('test info')
+  logger.warn('test warn')
+  logger.error('test error')
+  logger.fatal('test fatal')
+
+  await sleep(100)
```

Comment:
isn't there a way to make this more robust, iow not relying on an arbitrary number but actually wait for something to happen?

Reply:
I cannot think of a simple solution for this. We could have a FS watcher or a periodic poller of sorts which would get notified on file changes and then wait for it to be updated 6 times. If we get past that watcher/poller that wait would already have confirmed what we are about to test in the remainder of the code. If the file does not get updated 6 times we would have to wait for some kind of timeout to pass.

I don't see this as a better option. Do you have some better idea? The overhead of the current solution is 300ms.

Reply:
Instead of writing to a file, write to a custom stream and inspect the lines. https://github.com/pinojs/pino/issues/1274#issuecomment-997971329

Reply:
@jsumners, a question about the proposed solution. 
In [the code snippet](https://github.com/pinojs/pino/issues/1274#issuecomment-997971329) does the split function come from [split2](https://www.npmjs.com/package/split2) package?

In the opentelemetry transport the destination is being forwarded directly to the SonicBoom.
If the context.dest is a stream then SonicBoom cannot accept it.

`SonicBoom supports only file descriptors and files`

Reply:
> In [the code snippet](https://github.com/pinojs/pino/issues/1274#issuecomment-997971329) does the split function come from [split2](https://www.npmjs.com/package/split2) package?

I believe so, yes. Unfortunately, I no longer have access to that code to verify.

> In the opentelemetry transport the destination is being forwarded directly to the SonicBoom.
> If the context.dest is a stream then SonicBoom cannot accept it.

I missed that. I think the transport should be able to write to a stream. This is the baseline functionality of a transport -- https://getpino.io/#/docs/transports?id=v7-transports

Reply:
Shouldn't the options such as dest be serializable? The code is executed in a worker thread.

This is specific to v7 transports I think.

_Note here we've specified a module by package rather than by relative path. The options object we provide is serialized and injected into the transport worker thread, then passed to the module's exported function. This means that the options object can only contain types that are supported by the [Structured Clone Algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) which is used to (de)serializing objects between threads._
[V7 transports docs](https://github.com/pinojs/pino/blob/master/docs/transports.md#v7-transports)

Reply:
I don’t understand the question. Can you make it a bit clearer?

Reply:
To clarify:

Is it a requirement for a pino transport to accept a stream as a parameter? Like this:

```
const transport = pino.transport({
  target: 'pino-opentelemetry-transport',
  options: { destination: someWriteStream }
})

```
Or is it ok for it to just accept filenames [like in the example from the docs](https://github.com/pinojs/pino/blob/master/docs/transports.md#writing-a-transport)?

The OT-transport implementation does not accept a stream but instantiates a SonicBoom stream which writes to the specified destination.

Reply:
Who will pick this data up and collect it? I'm totally ok this being a file but then who will deliver it? There is no requirement for this to accept a stream.